src/main.rs

```diff
@@ -5,9 +5,36 @@ use vector::app::Application;
 
 #[cfg(unix)]
 fn main() {
-    // Configure our tracking allocator.
     #[cfg(feature = "allocation-tracing")]
-    vector::internal_telemetry::allocations::init_allocation_tracing();
+    {
+        use crate::vector::internal_telemetry::allocations::{
+            init_allocation_tracing, REPORTING_INTERVAL_MS, STARTUP, TRACK_ALLOCATIONS,
+        };
+        let opts = vector::cli::Opts::get_matches()
+            .map_err(|error| {
+                // Printing to stdout/err can itself fail; ignore it.
+                let _ = error.print();
+                exitcode::USAGE
+            })
+            .unwrap_or_else(|code| {
+                std::process::exit(code);
+            });
+        let allocation_tracing = opts.root.allocation_tracing;
+        REPORTING_INTERVAL_MS.store(
+            opts.root.allocation_tracing_reporting_interval_ms,
+            Ordering::Relaxed,
+        );
+        drop(opts);
+        use std::sync::atomic::Ordering;
+        STARTUP.store(false, Ordering::Relaxed);
+        if allocation_tracing {
+            // Configure our tracking allocator.
+            init_allocation_tracing();
+        } else {
+            // We assume nothing is allocated on the heap at this point.
+            TRACK_ALLOCATIONS.store(false, Ordering::Relaxed);
```

Comment:
I think we should make these `Release`, fitting with the other comments to make the loads `Acquire`, unless we can prove that the LLVM memory model, and the circumstances of where these loads/stores are happening (in `main`, no other threads, etc), provide a guarantee that no previous value can be read (i.e. reading `true` during load in the tracking allocator after actually having set it to `false`).